### DISCUSSION TITLE
Custom attribute on prim when importing as maya data

### DISCUSSION THREAD

**Discussion by jhultgre** [Autodesk/maya-usd#1290] [Q&A]
I'm trying to add a custom attribute to my usd prims with usd core.
Is there a way to have that show up in the attribute editor when I load the scene as maya data?
I see USD_kind showing up.

```
#usda 1.0
(
    defaultPrim = "group"
)

def Xform "group"
{
    over "obj1" (
        prepend references = @asset1.usd@
    )
    {
        custom string source_asset = "asset1"
        float3 xformOp:rotateXYZ = (0, 0, 0)
        float3 xformOp:scale = (1, 1, 1)
        double3 xformOp:translate = (0.2635994, -2.946043, -6.07903147)
        float3 xformOp:translate:pivot = (0, 0, 0)
        uniform token[] xformOpOrder = ["xformOp:translate", "xformOp:translate:pivot", "xformOp:rotateXYZ", "xformOp:scale", "!invert!xformOp:translate:pivot"]
    }

    over "obj2" (
        prepend references = @asset2.usd@
    )
    {
        custom string source_asset = "asset2"
        float3 xformOp:rotateXYZ = (0, 0, 0)
        float3 xformOp:scale = (1, 1, 1)
        double3 xformOp:translate = (0.2635994, -2.946043, -6.07903147)
        float3 xformOp:translate:pivot = (0, 0, 0)
        uniform token[] xformOpOrder = ["xformOp:translate", "xformOp:translate:pivot", "xformOp:rotateXYZ", "xformOp:scale", "!invert!xformOp:translate:pivot"]
    }
}
```

**Comment by womanyoyoyo-adsk** [comment]
Hi @jhultgre,

Currently we don’t have a way to import custom attributes from USD to Maya data. That’s why you don’t see them in the attribute editor.

We do have a way to bring in metadata on import through our import command MayaUSDImport with the flag, -metadata (which takes multi strings). Since in Maya we don’t have the notion of metadata, we add the prefix ‘USD_’ and import it as a Maya custom attribute. We then use this prefix to differentiate from attributes when exporting and export this info back as USD metadata. The -metadata flag is defaulted to ‘hidden, instanceable, kind’. This is why you are seeing ‘USD_kind’ showing up in the Attribute Editor as a custom attributes.

As for bringing in custom attributes on import, we do have a design for this. Here it is below. It would be an option as part of the ‘File Type Specific Options’ when selecting ‘USD Import’ on File – Import.

![importCustomAttributes](https://user-images.githubusercontent.com/49446314/113197646-e997ec80-9232-11eb-860b-4c3b9c44e47a.gif)

‘Include Attributes’ would be defaulted to ‘None’ as is the current behavior. In case our customers have an abundance of custom attributes, we think that defaulting this option to none is the safest. The second option is ‘All’, in which case we would import all custom attributes. The third option is ‘Custom’ where we let users specify which fields they’d like to import specifically. We would also have a flag for the MayaUSDImport command as well. The label ‘Include Attributes’ would have a tooltip educating the user of its function in more depth. 

What are your thoughts on this design? Would this solve your problem? Do you have any suggestions? Does the default of 'None' make sense? 

I want to add that we also have a very similar design for ‘Include Metadata’. This design clarifies what is happening under the hood with the ‘USD_’ prefix as I explained above. If you’re interested in this, can you please start another Discussion thread?

**Reply by jhultgre** [reply]
Is the a python flag for that? I have this wrapped up in a tool so the artists don't have to manage files they have synced.

**Reply by jhultgre** [reply]
I see it's in the other link.

**Reply by womanyoyoyo-adsk** [reply]
@jhultgre Did you find what you're looking for?

**Comment by womanyoyoyo-adsk** [comment]
@jhultgre, if your custom attributes on the prim were defined by an API schema, we have a flag for MayaUSDImport, -apiSchema, which takes multi-strings. See details here: https://github.com/Autodesk/maya-usd/blob/dev/plugin/pxr/doc/README.md. So I wasn't correct when I said we don’t have a way to import custom attributes from USD to Maya data.

**Reply by jhultgre** [reply]
I haven't tried making an API schema yet. I don't really know how to make one of those and I haven't really found a good tutorial for that.
